src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1724,283 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+            var tupleTypes = targetPlaceholder.Type.TupleElementTypes;
+            if (variables.Length != tupleTypes.Length)
+            {
+                Error(diagnostics, ErrorCode.ERR_DeconstructWrongCardinality, syntax, tupleTypes.Length, variables.Length);
+                return null;
+            }
 
-            TypeSymbol voidType = GetSpecialType(SpecialType.System_Void, diagnostics, node);
-            return new BoundDeconstructionAssignmentOperator(node, checkedVariables, typedRHS, deconstructMemberOpt: null, assignments: assignments, type: voidType);
+            return new BoundDeconstructionDeconstructStep(syntax, null, targetPlaceholder, tupleTypes.SelectAsArray((t, s) => new BoundDeconstructValuePlaceholder(s, t), syntax));
         }
 
-        private BoundExpression BindDeconstructWithDeconstruct(AssignmentExpressionSyntax node, DiagnosticBag diagnostics, BoundExpression boundRHS, ImmutableArray<BoundExpression> checkedVariables)
+        /// <summary>
+        /// This will generate and stack appropriate deconstruction and assignment steps for a non-tuple type.
+        /// Returns null if there was an error (if a suitable Deconstruct method was not found).
+        /// </summary>
+        static private BoundDeconstructionDeconstructStep MakeNonTupleDeconstructStep(
+                                                            BoundDeconstructValuePlaceholder targetPlaceholder,
+                                                            AssignmentExpressionSyntax syntax,
+                                                            DiagnosticBag diagnostics,
+                                                            ImmutableArray<DeconstructionVariable> variables,
+                                                            ArrayBuilder<BoundDeconstructionDeconstructStep> deconstructionSteps,
+                                                            ArrayBuilder<BoundDeconstructionAssignmentStep> assignmentSteps)
         {
             // symbol and parameters for Deconstruct
-            DiagnosticBag bag = new DiagnosticBag();
-            MethodSymbol deconstructMethod = FindDeconstruct(checkedVariables, boundRHS, node, bag);
-            if (!diagnostics.HasAnyErrors())
+            MethodSymbol deconstructMethod = FindDeconstruct(variables.Length, targetPlaceholder, syntax, diagnostics);
+
+            if ((object)deconstructMethod == null)
             {
-                diagnostics.AddRange(bag);
+                return null;
             }
 
-            if ((object)deconstructMethod == null)
+            return new BoundDeconstructionDeconstructStep(syntax, deconstructMethod, targetPlaceholder, deconstructMethod.Parameters.SelectAsArray((p, s) => new BoundDeconstructValuePlaceholder(s, p.Type), syntax));
+        }
+
+        private class DeconstructionVariable
+        {
+            public BoundExpression Single;
+            public ImmutableArray<DeconstructionVariable>? Nested;
```

Comment:
There is a type loader bug that would cause a struct S1 containing ImmutableArray<S1> be rejected as having a layout cycle (there is not real cycle though since IA wraps a ref array)
 CLR knows about this bug, but it is too risky area to fix right now.